### PR TITLE
flexible memory model - backend

### PR DESCRIPTION
**Pull request by Tantalus13A98B5F**

### CODE REVIEW

File: dev-clean/headers/llsc/state_imp.hpp [hunk 1]
```diff
@@ -4,47 +4,59 @@
 /* Memory, stack, and symbolic state representation */
 
 // Note (5/17): now using a byte-oriented layout
-template <class V>
+template <class V, class M>
 class PreMem {
```
Comment (Kraks) [inline dev-clean/headers/llsc/state_imp.hpp:8]: why PreMem needs this F-bounded-style polymorphism?

**Comment by Tantalus13A98B5F** [comment]
```
[info] TestPureCPSLLSC:
[info] - addTest
[info] - powerTest
[info] - globalTest
[info] - makeSymbolicTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "ureCPSLLSC_makeSymbolicTest"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - ptrPredTest
[info] - switchConcreteTest
[info] - truncTest
[info] - floatArithTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "erminat"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - arrayAccTest
[info] - arrayAccLocalTest
[info] - arrayGetSetTest
[info] - complexStructTest
[info] - structReturnLongTest
[info] - structAccessTest
[info] - structReturnTest
[info] - funptr
[info] - heapFunptr
[info] - branch1
[info] - branch2
[info] - branch3
[info] - switchSymTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "ureCPSLLSC_switchSymTest"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - mergeSortTest
[info] - bubbleSortTest
[info] - quickSortTest
[info] - kmp
[info] - k_of_st_arrays
[info] - binSearch
[info] - knapsackTest *** FAILED ***
[info]   1653 did not equal 1666 Unexpected path number (TestLLSC.scala:143)
[info] - mazeTest
[info] - mp1024Test
[info] - assertTest
[info] TestImpCPSLLSC:
[info] - addTest
[info] - powerTest
[info] - globalTest
[info] - makeSymbolicTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "mpCPSLLSC_makeSymbolicTest"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - ptrPredTest
[info] - switchConcreteTest
[info] - truncTest
[info] - floatArithTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "s"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - arrayAccTest
[info] - arrayAccLocalTest
[info] - arrayGetSetTest
[info] - complexStructTest
[info] - structReturnLongTest
[info] - structAccessTest
[info] - structReturnTest
[info] - funptr
[info] - heapFunptr
[info] - branch1
[info] - branch2
[info] - branch3
[info] - switchSymTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "mpCPSLLSC_switchSymTest"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - mergeSortTest
[info] - bubbleSortTest
[info] - quickSortTest
[info] - kmp
[info] - k_of_st_arrays
[info] - binSearch
[info] - knapsackTest *** FAILED ***
[info]   1653 did not equal 1666 Unexpected path number (TestLLSC.scala:143)
[info] - mazeTest
[info] - mp1024Test
[info] - assertTest
[info] TestPureLLSC:
[info] - addTest
[info] - powerTest
[info] - globalTest
[info] - makeSymbolicTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "ureLLSC_makeSymbolicTest"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - ptrPredTest
[info] - switchConcreteTest
[info] - truncTest
[info] - floatArithTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "erminat"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
TestResult(PureLLSC,openTestFail,1.87E-4,6.28E-4,0.5,1,0,1,0)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - arrayAccTest
[info] - arrayAccLocalTest
[info] - arrayGetSetTest
[info] - complexStructTest
[info] - structReturnLongTest
[info] - structAccessTest
[info] - structReturnTest
[info] - funptr
[info] - heapFunptr
[info] - varArgInt
[info] - branch1
[info] - branch2
[info] - branch3
[info] - switchSymTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "ureLLSC_switchSymTest"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - mergeSortTest
[info] - bubbleSortTest
[info] - quickSortTest
[info] - kmp
[info] - k_of_st_arrays
[info] - binSearch
[info] - knapsackTest *** FAILED ***
[info]   1653 did not equal 1666 Unexpected path number (TestLLSC.scala:143)
[info] - mazeTest
[info] - mp1024Test
[info] - assertTest
[info] - openTestSucc
[info] - openTestFail
[info] TestPureCPSLLSC_Z3:
[info] - addTest
[info] - powerTest
[info] - globalTest
[info] - makeSymbolicTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "ureCPSLLSC_Z3_makeSymbolicTest"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - ptrPredTest
[info] - switchConcreteTest
[info] - truncTest
[info] - floatArithTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "erminat"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - arrayAccTest
[info] - arrayAccLocalTest
[info] - arrayGetSetTest
[info] - complexStructTest
[info] - structReturnLongTest
[info] - structAccessTest
[info] - structReturnTest
[info] - funptr
[info] - heapFunptr
[info] - branch1
[info] - branch2
[info] - branch3
[info] - switchSymTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "ureCPSLLSC_Z3_switchSymTest"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - mergeSortTest
[info] - bubbleSortTest
[info] - quickSortTest
[info] - kmp
[info] - k_of_st_arrays
[info] - binSearch
[info] - knapsackTest *** FAILED ***
[info]   1653 did not equal 1666 Unexpected path number (TestLLSC.scala:143)
[info] - mazeTest
[info] - mp1024Test
[info] - assertTest
[info] TestImpLLSC:
[info] - addTest
[info] - powerTest
[info] - globalTest
[info] - makeSymbolicTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "mpLLSC_makeSymbolicTest"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - ptrPredTest
[info] - switchConcreteTest
[info] - truncTest
[info] - floatArithTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "erminat"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - arrayAccTest
[info] - arrayAccLocalTest
[info] - arrayGetSetTest
[info] - complexStructTest
[info] - structReturnLongTest
[info] - structAccessTest
[info] - structReturnTest
[info] - funptr
[info] - heapFunptr
[info] - branch1
[info] - branch2
[info] - branch3
[info] - switchSymTest *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "mpLLSC_switchSymTest"
[info]   at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
[info]   at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
[info]   at java.lang.Double.parseDouble(Double.java:538)
[info]   at scala.collection.immutable.StringLike.toDouble(StringLike.scala:321)
[info]   at scala.collection.immutable.StringLike.toDouble$(StringLike.scala:321)
[info]   at scala.collection.immutable.StringOps.toDouble(StringOps.scala:33)
[info]   at sai.llsc.TestLLSC.parseOutput(TestLLSC.scala:114)
[info]   at sai.llsc.TestLLSC.$anonfun$testLLSC$1(TestLLSC.scala:140)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - mergeSortTest
[info] - bubbleSortTest
[info] - quickSortTest
[info] - kmp
[info] - k_of_st_arrays
[info] - binSearch
[info] - knapsackTest *** FAILED ***
[info]   1653 did not equal 1666 Unexpected path number (TestLLSC.scala:143)
[info] - mazeTest
[info] - mp1024Test
[info] - assertTest
[info] Run completed in 1 hour, 7 minutes, 17 seconds.
[info] Total number of tests run: 160
[info] Suites: completed 6, aborted 0
[info] Tests: succeeded 140, failed 20, canceled 0, ignored 0, pending 0
[info] *** 20 TESTS FAILED ***
```

**Comment by Tantalus13A98B5F** [comment]
seems to have come to a milestone... passed all tests except FS- and Z3-related ones.

**Comment by Tantalus13A98B5F** [comment]
Update: now Z3 is supported

@PROgram52bc Would you like to have a look at the FS tests on this branch?

**Pull request merged**